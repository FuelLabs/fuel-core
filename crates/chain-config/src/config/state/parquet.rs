pub mod decode;
pub mod encode;

#[cfg(test)]
mod tests {
    use bytes::Bytes;
    use itertools::Itertools;
    use parquet::{
        basic::ZstdLevel,
        file::reader::{
            ChunkReader,
            Length,
        },
    };
    use rand::{
        rngs::StdRng,
        Rng,
        SeedableRng,
    };
    use std::{
        io::{
            Cursor,
            Read,
            Seek,
            SeekFrom,
        },
        iter::repeat_with,
        sync::{
            atomic::AtomicU64,
            Arc,
            Mutex,
        },
    };

    use super::{
        decode::*,
        encode::*,
    };

    #[derive(Clone)]
    struct TrackingReader {
        source: Arc<Mutex<Cursor<Vec<u8>>>>,
        read_bytes: Arc<AtomicU64>,
    }

    impl TrackingReader {
        fn new(source: Vec<u8>) -> Self {
            Self {
                source: Arc::new(Mutex::new(Cursor::new(source))),
                read_bytes: Default::default(),
            }
        }
    }

    impl Seek for TrackingReader {
        fn seek(&mut self, pos: SeekFrom) -> std::io::Result<u64> {
            self.source.lock().unwrap().seek(pos)
        }
    }

    impl Read for TrackingReader {
        fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
            let amount = self.source.lock().unwrap().read(buf)?;
            self.read_bytes
                .fetch_add(amount as u64, std::sync::atomic::Ordering::SeqCst);
            Ok(amount)
        }
    }

    impl Length for TrackingReader {
        fn len(&self) -> u64 {
            let mut source = self.source.lock().unwrap();

            let current = source.stream_position().unwrap();

            source.seek(SeekFrom::Start(0)).unwrap();
            let len = source.seek(SeekFrom::End(0)).unwrap();

            source.seek(SeekFrom::Start(current)).unwrap();
            len
        }
    }

    impl ChunkReader for TrackingReader {
        type T = TrackingReader;

        fn get_read(&self, start: u64) -> parquet::errors::Result<Self::T> {
            let mut new_reader = self.clone();
            new_reader.seek(SeekFrom::Start(start))?;

            Ok(new_reader)
        }

        fn get_bytes(&self, start: u64, length: usize) -> parquet::errors::Result<Bytes> {
            let mut buf = Vec::with_capacity(length);
            let reader = self.get_read(start)?;
            reader.take(length as u64).read_to_end(&mut buf)?;

            Ok(Bytes::from(buf))
        }
    }

    #[test]
    fn can_skip_groups_without_reading_whole_file() {
        // given
        let mut buffer = vec![];
        let mut encoder = Encoder::new(
            &mut buffer,
            parquet::basic::Compression::ZSTD(ZstdLevel::try_new(1).unwrap()),
        )
        .unwrap();
        let mut rng = StdRng::seed_from_u64(0);

        let big_group = repeat_with(|| rng.gen::<[u8; 32]>().to_vec())
            .take(1000)
            .collect_vec();
        encoder.write(big_group).unwrap();

        let small_group = vec![rng.gen::<[u8; 32]>().to_vec()];
        encoder.write(small_group).unwrap();
        encoder.close().unwrap();
        let total_size = buffer.len();

        let bytes = TrackingReader::new(buffer);
        let bytes_read = bytes.read_bytes.clone();

        let mut decoder = Decoder::new(bytes).unwrap();

        // when
        let _: Vec<_> = decoder.nth(1).unwrap().unwrap();

        // then
        let actually_read = bytes_read.load(std::sync::atomic::Ordering::SeqCst);

        assert_eq!(total_size, 36930);
        assert_eq!(actually_read, 509);
    }
}
