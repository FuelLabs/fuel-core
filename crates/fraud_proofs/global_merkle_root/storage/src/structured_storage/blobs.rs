use crate::{
    column::TableColumn,
    merkle::MerkleizedTableColumn,
};
use fuel_core_storage::tables::BlobData;

impl MerkleizedTableColumn for BlobData {
    fn table_column() -> TableColumn {
        TableColumn::Blobs
    }
}
