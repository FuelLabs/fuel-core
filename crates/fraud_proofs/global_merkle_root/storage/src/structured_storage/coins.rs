use crate::{
    column::TableColumn,
    merkle::MerkleizedTableColumn,
};
use fuel_core_storage::tables::Coins;

impl MerkleizedTableColumn for Coins {
    fn table_column() -> TableColumn {
        TableColumn::Coins
    }
}
