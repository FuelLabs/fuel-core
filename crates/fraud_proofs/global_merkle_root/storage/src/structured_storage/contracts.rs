use crate::{
    column::TableColumns,
    merkle::MerklizedTableColumn,
};
use fuel_core_storage::tables::{
    ContractsLatestUtxo,
    ContractsRawCode,
};

impl MerklizedTableColumn for ContractsRawCode {
    fn table_column() -> TableColumns {
        TableColumns::ContractsRawCode
    }
}

impl MerklizedTableColumn for ContractsLatestUtxo {
    fn table_column() -> TableColumns {
        TableColumns::ContractsLatestUtxo
    }
}
