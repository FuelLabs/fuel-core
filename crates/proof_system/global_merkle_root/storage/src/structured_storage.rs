mod blobs;
mod coins;
mod contracts;
mod messages;
mod transactions;
mod upgrades;
