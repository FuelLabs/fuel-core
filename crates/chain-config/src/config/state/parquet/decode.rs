use anyhow::{
    anyhow,
    Context,
};
use parquet::{
    data_type::AsBytes,
    file::{
        reader::{
            ChunkReader,
            FileReader,
        },
        serialized_reader::SerializedFileReader,
    },
    record::RowAccessor,
};

pub struct Decoder<R: ChunkReader> {
    data_source: SerializedFileReader<R>,
    group_index: usize,
}

impl<R> Decoder<R>
where
    R: ChunkReader + 'static,
{
    pub fn num_groups(&self) -> usize {
        self.data_source.num_row_groups()
    }

    fn current_group(&self) -> anyhow::Result<Vec<Vec<u8>>> {
        let data = self
            .data_source
            .get_row_group(self.group_index)?
            .get_row_iter(None)?
            .map(|result| {
                result.map_err(|e| anyhow!(e)).and_then(|row| {
                    const FIELD_IDX: usize = 0;
                    Ok(row
                        .get_bytes(FIELD_IDX)
                        .context("While decoding postcard bytes")?
                        .as_bytes()
                        .to_vec())
                })
            })
            .collect::<Result<Vec<_>, _>>()?;

        Ok(data)
    }
}

impl<R> Iterator for Decoder<R>
where
    R: ChunkReader + 'static,
{
    type Item = anyhow::Result<Vec<Vec<u8>>>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.group_index >= self.data_source.metadata().num_row_groups() {
            return None;
        }

        let group = self.current_group();
        self.group_index = self.group_index.saturating_add(1);

        Some(group)
    }

    fn nth(&mut self, n: usize) -> Option<Self::Item> {
        self.group_index = self.group_index.saturating_add(n);
        self.next()
    }
}

impl<R: ChunkReader + 'static> Decoder<R> {
    pub fn new(reader: R) -> anyhow::Result<Self> {
        Ok(Self {
            data_source: SerializedFileReader::new(reader)?,
            group_index: 0,
        })
    }
}
