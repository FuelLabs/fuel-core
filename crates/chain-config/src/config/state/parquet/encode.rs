use std::{
    io::Write,
    sync::Arc,
};

use itertools::Itertools;
use parquet::{
    basic::{
        Compression,
        Repetition,
    },
    file::{
        properties::WriterProperties,
        writer::SerializedFileWriter,
    },
};

use parquet::{
    data_type::ByteArrayType,
    schema::types::Type,
};

pub struct Encoder<W>
where
    W: Write,
{
    writer: SerializedFileWriter<W>,
}

impl<W> Encoder<W>
where
    W: Write + Send,
{
    pub fn new(writer: W, compression: Compression) -> anyhow::Result<Self> {
        let writer = SerializedFileWriter::new(
            writer,
            Arc::new(Self::single_element_schema()),
            Arc::new(
                WriterProperties::builder()
                    .set_compression(compression)
                    .build(),
            ),
        )?;

        Ok(Self { writer })
    }

    fn single_element_schema() -> Type {
        let data =
            Type::primitive_type_builder("data", ::parquet::basic::Type::BYTE_ARRAY)
                .with_repetition(Repetition::REQUIRED)
                .build()
                .expect("This is a valid schema");

        Type::group_type_builder("unimportant")
            .with_fields(vec![Arc::new(data)])
            .build()
            .expect("This is a valid schema")
    }
}

impl<W> Encoder<W>
where
    W: Write + Send,
{
    pub fn write(&mut self, elements: Vec<Vec<u8>>) -> anyhow::Result<()> {
        let mut group = self.writer.next_row_group()?;
        let mut column = group
            .next_column()?
            .ok_or_else(|| anyhow::anyhow!("Missing column. Check the schema!"))?;

        let values = elements.into_iter().map(Into::into).collect_vec();
        column
            .typed::<ByteArrayType>()
            .write_batch(&values, None, None)?;

        column.close()?;
        group.close()?;
        Ok(())
    }

    pub fn close(self) -> anyhow::Result<()> {
        self.writer.close()?;
        Ok(())
    }
}
