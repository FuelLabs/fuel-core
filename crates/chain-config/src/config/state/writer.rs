use crate::{
    config::table_entry::TableEntry,
    AddTable,
    ChainConfig,
    LastBlockConfig,
    SnapshotMetadata,
    StateConfigBuilder,
    TableEncoding,
};
use fuel_core_storage::structured_storage::TableWithBlueprint;
use std::path::PathBuf;

#[cfg(feature = "parquet")]
use super::parquet;

enum EncoderType {
    Json {
        builder: StateConfigBuilder,
    },
    #[cfg(feature = "parquet")]
    Parquet {
        compression: ZstdCompressionLevel,
        table_encoders: TableEncoders,
    },
}

pub struct SnapshotWriter {
    dir: PathBuf,
    encoder: EncoderType,
}

#[allow(dead_code)]
#[derive(
    Debug,
    Clone,
    Copy,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    serde::Serialize,
    serde::Deserialize,
)]
#[cfg(feature = "parquet")]
#[cfg_attr(test, derive(strum::EnumIter))]
pub enum ZstdCompressionLevel {
    Uncompressed,
    Level1,
    Level2,
    Level3,
    Level4,
    Level5,
    Level6,
    Level7,
    Level8,
    Level9,
    Level10,
    Level11,
    Level12,
    Level13,
    Level14,
    Level15,
    Level16,
    Level17,
    Level18,
    Level19,
    Level20,
    Level21,
    Max,
}

#[cfg(feature = "parquet")]
impl TryFrom<u8> for ZstdCompressionLevel {
    type Error = anyhow::Error;
    fn try_from(value: u8) -> Result<Self, Self::Error> {
        match value {
            0 => Ok(Self::Uncompressed),
            1 => Ok(Self::Level1),
            2 => Ok(Self::Level2),
            3 => Ok(Self::Level3),
            4 => Ok(Self::Level4),
            5 => Ok(Self::Level5),
            6 => Ok(Self::Level6),
            7 => Ok(Self::Level7),
            8 => Ok(Self::Level8),
            9 => Ok(Self::Level9),
            10 => Ok(Self::Level10),
            11 => Ok(Self::Level11),
            12 => Ok(Self::Level12),
            13 => Ok(Self::Level13),
            14 => Ok(Self::Level14),
            15 => Ok(Self::Level15),
            16 => Ok(Self::Level16),
            17 => Ok(Self::Level17),
            18 => Ok(Self::Level18),
            19 => Ok(Self::Level19),
            20 => Ok(Self::Level20),
            21 => Ok(Self::Level21),
            22 => Ok(Self::Max),
            _ => {
                anyhow::bail!("Compression level {value} outside of allowed range 0..=22")
            }
        }
    }
}

#[cfg(feature = "parquet")]
impl From<ZstdCompressionLevel> for u8 {
    fn from(value: ZstdCompressionLevel) -> Self {
        match value {
            ZstdCompressionLevel::Uncompressed => 0,
            ZstdCompressionLevel::Level1 => 1,
            ZstdCompressionLevel::Level2 => 2,
            ZstdCompressionLevel::Level3 => 3,
            ZstdCompressionLevel::Level4 => 4,
            ZstdCompressionLevel::Level5 => 5,
            ZstdCompressionLevel::Level6 => 6,
            ZstdCompressionLevel::Level7 => 7,
            ZstdCompressionLevel::Level8 => 8,
            ZstdCompressionLevel::Level9 => 9,
            ZstdCompressionLevel::Level10 => 10,
            ZstdCompressionLevel::Level11 => 11,
            ZstdCompressionLevel::Level12 => 12,
            ZstdCompressionLevel::Level13 => 13,
            ZstdCompressionLevel::Level14 => 14,
            ZstdCompressionLevel::Level15 => 15,
            ZstdCompressionLevel::Level16 => 16,
            ZstdCompressionLevel::Level17 => 17,
            ZstdCompressionLevel::Level18 => 18,
            ZstdCompressionLevel::Level19 => 19,
            ZstdCompressionLevel::Level20 => 20,
            ZstdCompressionLevel::Level21 => 21,
            ZstdCompressionLevel::Max => 22,
        }
    }
}

#[cfg(feature = "parquet")]
impl From<ZstdCompressionLevel> for ::parquet::basic::Compression {
    fn from(value: ZstdCompressionLevel) -> Self {
        if let ZstdCompressionLevel::Uncompressed = value {
            Self::UNCOMPRESSED
        } else {
            let level = i32::from(u8::from(value));
            let level = ::parquet::basic::ZstdLevel::try_new(level)
                .expect("our range to mimic the parquet zstd range");
            Self::ZSTD(level)
        }
    }
}

#[derive(Debug, Clone)]
enum FragmentData {
    Json {
        builder: StateConfigBuilder,
    },
    #[cfg(feature = "parquet")]
    Parquet {
        tables: std::collections::HashMap<String, PathBuf>,
        compression: ZstdCompressionLevel,
    },
}
impl FragmentData {
    fn merge(mut self, data: FragmentData) -> anyhow::Result<Self> {
        match (&mut self, data) {
            (
                FragmentData::Json { builder, .. },
                FragmentData::Json {
                    builder: other_builder,
                    ..
                },
            ) => {
                builder.merge(other_builder);
            }
            #[cfg(feature = "parquet")]
            (
                FragmentData::Parquet {
                    tables,
                    compression,
                },
                FragmentData::Parquet {
                    tables: their_tables,
                    compression: their_compression,
                },
            ) => {
                tables.extend(their_tables);
                anyhow::ensure!(*compression == their_compression, "Fragments use different compressions.")
            }
            #[cfg(feature="parquet")]
            (a,b) => anyhow::bail!("Fragments don't have the same encoding and cannot be merged. Fragments: {a:?} and {b:?}"),
        };

        Ok(self)
    }
}

#[derive(Debug, Clone)]
pub struct SnapshotFragment {
    dir: PathBuf,
    data: FragmentData,
}

impl SnapshotFragment {
    pub fn merge(mut self, fragment: Self) -> anyhow::Result<Self> {
        self.data = self.data.merge(fragment.data)?;
        Ok(self)
    }

    pub fn finalize(
        self,
        latest_block_config: Option<LastBlockConfig>,
        chain_config: &ChainConfig,
    ) -> anyhow::Result<SnapshotMetadata> {
        let table_encoding = match self.data {
            FragmentData::Json { builder } => {
                let state_config = builder.build(latest_block_config)?;
                std::fs::create_dir_all(&self.dir)?;
                let state_file_path = self.dir.join("state_config.json");
                let file = std::fs::File::create(&state_file_path)?;
                serde_json::to_writer_pretty(file, &state_config)?;

                TableEncoding::Json {
                    filepath: state_file_path,
                }
            }
            #[cfg(feature = "parquet")]
            FragmentData::Parquet {
                tables,
                compression,
            } => {
                let latest_block_config_path =
                    self.dir.join("latest_block_config.parquet");
                SnapshotWriter::write_single_el_parquet(
                    &latest_block_config_path,
                    latest_block_config,
                    compression,
                )?;

                TableEncoding::Parquet {
                    tables,
                    latest_block_config_path,
                }
            }
        };

        SnapshotWriter::write_chain_config_and_metadata(
            &self.dir,
            chain_config,
            table_encoding,
        )
    }
}

impl SnapshotWriter {
    const CHAIN_CONFIG_FILENAME: &'static str = "chain_config.json";
    pub fn json(dir: impl Into<PathBuf>) -> Self {
        Self {
            encoder: EncoderType::Json {
                builder: StateConfigBuilder::default(),
            },
            dir: dir.into(),
        }
    }

    #[cfg(feature = "parquet")]
    pub fn parquet(
        dir: impl Into<::std::path::PathBuf>,
        compression_level: ZstdCompressionLevel,
    ) -> anyhow::Result<Self> {
        let dir = dir.into();
        std::fs::create_dir_all(&dir)?;
        Ok(Self {
            encoder: EncoderType::Parquet {
                table_encoders: TableEncoders::new(dir.clone(), compression_level),
                compression: compression_level,
            },
            dir,
        })
    }

    #[cfg(feature = "test-helpers")]
    pub fn write_state_config(
        mut self,
        state_config: crate::StateConfig,
        chain_config: &ChainConfig,
    ) -> anyhow::Result<SnapshotMetadata> {
        use fuel_core_storage::tables::{
            Coins,
            ContractsAssets,
            ContractsLatestUtxo,
            ContractsRawCode,
            ContractsState,
            Messages,
        };
        use fuel_core_types::fuel_vm::BlobData;

        use crate::AsTable;

        self.write::<Coins>(state_config.as_table())?;
        self.write::<Messages>(state_config.as_table())?;
        self.write::<BlobData>(state_config.as_table())?;
        self.write::<ContractsRawCode>(state_config.as_table())?;
        self.write::<ContractsLatestUtxo>(state_config.as_table())?;
        self.write::<ContractsState>(state_config.as_table())?;
        self.write::<ContractsAssets>(state_config.as_table())?;
        self.close(state_config.last_block, chain_config)
    }

    pub fn write<T>(&mut self, elements: Vec<TableEntry<T>>) -> anyhow::Result<()>
    where
        T: TableWithBlueprint,
        TableEntry<T>: serde::Serialize,
        StateConfigBuilder: AddTable<T>,
    {
        match &mut self.encoder {
            EncoderType::Json { builder, .. } => {
                builder.add(elements);
                Ok(())
            }
            #[cfg(feature = "parquet")]
            EncoderType::Parquet { table_encoders, .. } => {
                table_encoders.encoder::<T>()?.write::<T>(elements)
            }
        }
    }

    #[cfg(feature = "parquet")]
    fn write_single_el_parquet(
        path: &std::path::Path,
        data: impl serde::Serialize,
        compression: ZstdCompressionLevel,
    ) -> anyhow::Result<()> {
        let mut encoder = parquet::encode::Encoder::new(
            std::fs::File::create(path)?,
            compression.into(),
        )?;
        encoder.write(vec![postcard::to_stdvec(&data)?])?;
        encoder.close()
    }

    pub fn close(
        self,
        latest_block_config: Option<LastBlockConfig>,
        chain_config: &ChainConfig,
    ) -> anyhow::Result<SnapshotMetadata> {
        self.partial_close()?
            .finalize(latest_block_config, chain_config)
    }

    fn write_chain_config_and_metadata(
        dir: &std::path::Path,
        chain_config: &ChainConfig,
        table_encoding: TableEncoding,
    ) -> anyhow::Result<SnapshotMetadata> {
        let chain_config_path = dir.join(Self::CHAIN_CONFIG_FILENAME);
        chain_config.write(&chain_config_path)?;

        let metadata = SnapshotMetadata {
            chain_config: chain_config_path,
            table_encoding,
        };
        metadata.clone().write(dir)?;
        Ok(metadata)
    }

    pub fn partial_close(self) -> anyhow::Result<SnapshotFragment> {
        let data = match self.encoder {
            EncoderType::Json { builder } => FragmentData::Json { builder },
            #[cfg(feature = "parquet")]
            EncoderType::Parquet {
                table_encoders,
                compression,
                ..
            } => {
                let tables = table_encoders.close()?;
                FragmentData::Parquet {
                    tables,
                    compression,
                }
            }
        };
        let snapshot_fragment = SnapshotFragment {
            dir: self.dir,
            data,
        };
        Ok(snapshot_fragment)
    }
}

#[cfg(feature = "parquet")]
struct PostcardParquetEncoder {
    path: PathBuf,
    encoder: parquet::encode::Encoder<std::fs::File>,
}

#[cfg(feature = "parquet")]
impl PostcardParquetEncoder {
    pub fn new(path: PathBuf, encoder: parquet::encode::Encoder<std::fs::File>) -> Self {
        Self { path, encoder }
    }

    fn write<T>(&mut self, elements: Vec<TableEntry<T>>) -> anyhow::Result<()>
    where
        T: fuel_core_storage::Mappable,
        TableEntry<T>: serde::Serialize,
    {
        use itertools::Itertools;
        let encoded: Vec<_> = elements
            .into_iter()
            .map(|entry| postcard::to_stdvec(&entry))
            .try_collect()?;
        self.encoder.write(encoded)
    }
}

#[cfg(feature = "parquet")]
struct TableEncoders {
    dir: PathBuf,
    compression: ZstdCompressionLevel,
    encoders: std::collections::HashMap<String, PostcardParquetEncoder>,
}

#[cfg(feature = "parquet")]
impl TableEncoders {
    fn new(dir: PathBuf, compression: ZstdCompressionLevel) -> Self {
        Self {
            dir,
            compression,
            encoders: Default::default(),
        }
    }

    fn encoder<T: fuel_core_storage::structured_storage::TableWithBlueprint>(
        &mut self,
    ) -> anyhow::Result<&mut PostcardParquetEncoder> {
        use fuel_core_storage::kv_store::StorageColumn;

        let name = StorageColumn::name(&T::column()).to_string();

        let encoder = match self.encoders.entry(name) {
            std::collections::hash_map::Entry::Occupied(encoder) => encoder.into_mut(),
            std::collections::hash_map::Entry::Vacant(vacant) => {
                let name = vacant.key();
                let file_path = self.dir.join(format!("{name}.parquet"));
                let file = std::fs::File::create(&file_path)?;
                let encoder = PostcardParquetEncoder::new(
                    file_path,
                    parquet::encode::Encoder::new(file, self.compression.into())?,
                );
                vacant.insert(encoder)
            }
        };

        Ok(encoder)
    }

    fn close(self) -> anyhow::Result<std::collections::HashMap<String, PathBuf>> {
        let mut files = std::collections::HashMap::new();
        for (file, encoder) in self.encoders {
            encoder.encoder.close()?;
            files.insert(file, encoder.path);
        }
        Ok(files)
    }
}

#[cfg(test)]
mod tests {
    use std::path::Path;

    use fuel_core_storage::{
        kv_store::StorageColumn,
        structured_storage::TableWithBlueprint,
        tables::{
            Coins,
            ContractsAssets,
            ContractsLatestUtxo,
            ContractsRawCode,
            ContractsState,
            Messages,
        },
    };
    use rand::{
        rngs::StdRng,
        SeedableRng,
    };

    use crate::StateConfig;

    use super::*;

    #[test]
    fn can_roundtrip_compression_level() {
        use strum::IntoEnumIterator;

        for level in crate::ZstdCompressionLevel::iter() {
            let u8_level = u8::from(level);
            let roundtrip = ZstdCompressionLevel::try_from(u8_level).unwrap();
            assert_eq!(level, roundtrip);
        }
    }

    #[test]
    fn parquet_encoder_encodes_tables_in_expected_files() {
        fn file_created_and_present_in_metadata<T>()
        where
            T: TableWithBlueprint,
            T::OwnedKey: serde::Serialize,
            T::OwnedValue: serde::Serialize,
            StateConfigBuilder: AddTable<T>,
        {
            // given
            use pretty_assertions::assert_eq;

            let dir = tempfile::tempdir().unwrap();
            let mut writer =
                SnapshotWriter::parquet(dir.path(), ZstdCompressionLevel::Uncompressed)
                    .unwrap();

            // when
            writer.write::<T>(vec![]).unwrap();
            let snapshot = writer.close(None, &ChainConfig::local_testnet()).unwrap();

            // then
            assert!(snapshot.chain_config.exists());
            let TableEncoding::Parquet { tables, .. } = snapshot.table_encoding else {
                panic!("Expected parquet encoding")
            };
            assert_eq!(tables.len(), 1, "Expected single table");
            let (table_name, path) = tables.into_iter().next().unwrap();

            assert_eq!(table_name, T::column().name());
            assert!(dir.path().join(path).exists());
        }

        file_created_and_present_in_metadata::<Coins>();
        file_created_and_present_in_metadata::<Messages>();
        file_created_and_present_in_metadata::<ContractsRawCode>();
        file_created_and_present_in_metadata::<ContractsLatestUtxo>();
        file_created_and_present_in_metadata::<ContractsState>();
        file_created_and_present_in_metadata::<ContractsAssets>();
    }

    #[test]
    fn all_compressions_are_valid() {
        use ::parquet::basic::Compression;
        use strum::IntoEnumIterator;
        for level in ZstdCompressionLevel::iter() {
            let _ = Compression::from(level);
        }
    }

    #[test]
    fn json_snapshot_is_human_readable() {
        // given
        use crate::Randomize;
        let dir = tempfile::tempdir().unwrap();
        let writer = SnapshotWriter::json(dir.path());
        let mut rng = StdRng::from_seed([0; 32]);
        let state = StateConfig::randomize(&mut rng);

        // when
        let snapshot = writer
            .write_state_config(state, &ChainConfig::local_testnet())
            .unwrap();

        // then
        let TableEncoding::Json { filepath } = snapshot.table_encoding else {
            panic!("Expected json encoding")
        };
        let encoded_json = std::fs::read_to_string(filepath).unwrap();

        insta::assert_snapshot!(encoded_json);
    }

    fn given_parquet_writer(path: &Path) -> SnapshotWriter {
        SnapshotWriter::parquet(path, ZstdCompressionLevel::Uncompressed).unwrap()
    }

    fn given_json_writer(path: &Path) -> SnapshotWriter {
        SnapshotWriter::json(path)
    }

    #[test_case::test_case(given_parquet_writer)]
    #[test_case::test_case(given_json_writer)]
    fn can_partially_close_without_chain_and_block_height(
        writer: impl Fn(&Path) -> SnapshotWriter + Copy,
    ) {
        // given
        let dir = tempfile::tempdir().unwrap();
        let writer = writer(dir.path());

        // when
        let result = writer.partial_close();

        // then
        assert!(result.is_ok());
    }

    #[test]
    fn merging_json_and_parquet_fragments_fails() {
        // given
        let dir = tempfile::tempdir().unwrap();
        let json_writer = super::SnapshotWriter::json(dir.path());
        let parquet_writer = super::SnapshotWriter::parquet(
            dir.path(),
            super::ZstdCompressionLevel::Uncompressed,
        );

        let json_fragment = json_writer.partial_close().unwrap();
        let parquet_fragment = parquet_writer.unwrap().partial_close().unwrap();

        {
            // when
            let result = json_fragment.clone().merge(parquet_fragment.clone());

            // when
            let err = result.unwrap_err();
            assert!(err.to_string().contains(
                "Fragments don't have the same encoding and cannot be merged."
            ));
        }
        {
            // when
            let result = parquet_fragment.merge(json_fragment);

            // when
            let err = result.unwrap_err();
            assert!(err.to_string().contains(
                "Fragments don't have the same encoding and cannot be merged."
            ));
        }
    }

    // It is enough just for the test to compile
    #[test]
    #[ignore]
    fn fragment_must_be_send_sync() {
        fn _assert_send<T: Send + Sync>() {}
        _assert_send::<super::SnapshotFragment>();
    }
}
