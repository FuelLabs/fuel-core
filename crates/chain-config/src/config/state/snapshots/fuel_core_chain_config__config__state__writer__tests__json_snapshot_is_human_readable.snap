---
source: crates/chain-config/src/config/state/writer.rs
expression: encoded_json
---
{
  "coins": [
    {
      "tx_id": "9b07815f04497e2e05d22cac3aa061410b20868cc619154c42a1c61be9902717",
      "output_index": 3250,
      "tx_pointer_block_height": 1405935179,
      "tx_pointer_tx_idx": 35726,
      "owner": "c168fe4896f5b70ccb10f1a705e061d0fdfa189618d28b0d44efef92cbf44b58",
      "amount": 16206947779870322207,
      "asset_id": "222c10916b17a369b4da039d075952b58036f2a7b561446592403c672f0dbe12"
    },
    {
      "tx_id": "9f1e39113f9f6164ea2867f8deccf3c3c518f146de9554f03f27a83d5b1185f3",
      "output_index": 10142,
      "tx_pointer_block_height": 476225137,
      "tx_pointer_tx_idx": 52207,
      "owner": "357e2bd9e13962516b0a41403276437b25f0ffbe897cf476c0777e9d246ff397",
      "amount": 901131989421222986,
      "asset_id": "f1b350aa16c42ea0d7117b30c8c8436ce7a9254b30d253e4567ccafa5f36ce84"
    }
  ],
  "messages": [
    {
      "sender": "c80aa8bc9be64e0d5ae796884274aef3005ae6733809c8ec1d5b84dd6289e193",
      "recipient": "b9f88de4a994358c9f856135236c3e75a925e1c77ac34127c8baefbea0939f15",
      "nonce": "24712b4d37a020b28cdd32138ae34d22e6397af3a9a71e4efa1fbad16cb0cfaf",
      "amount": 18009871113071078878,
      "data": "2884858e83fa4146c34e4d5ca8f7f7dfaa083b1c11af4b844ff94b3fbef6e36b",
      "da_height": 14407604056361705041
    },
    {
      "sender": "a3ad4c5856fa686464524a876b463d1297603568c40e814d9d5396d23087a0fd",
      "recipient": "641e91e0e00df6c012cded9ef9ba5e5bf042464b5ca6488b897c005b2e17f00f",
      "nonce": "c0bc8e7ed3f2ab59d607b0e3586bda1cda4e42b4145b26fa35ea19d258b113d8",
      "amount": 8710761994859176931,
      "data": "2bca2aa612b304ece5b25988818dd7234e049913233eb918c11638af89d575be",
      "da_height": 13053614869271975353
    }
  ],
  "blobs": [
    {
      "blob_id": "8acdb0431a136a5c0c100271704144e34aebe82cb426ced41461224e7798bceb",
      "payload": "d32d676c0ac47af4243e53915a314986c779af574aaf987302fc76b2e205b24e6e2b"
    },
    {
      "blob_id": "fb57963e95792ded6eed1ddbf40329cc6bdcb91decb9042aba6385216d8c07e2",
      "payload": "cb83691e8057f9e3ee966b281b5843029c002ea90022143aa9a8711d331894e545abb4a5ce6e6ae85ffb107712e030ddbe3a50f6a93ec455ff76b489e0f68a8a47ee42"
    }
  ],
  "contracts": [
    {
      "contract_id": "6b97a1efc738a7508436290b7495f2fed59d0454463c189bb56cf0996c6acb1a",
      "code": "48d6696c30631735193aea95b8f9082c10620c0d6e3fc9596b4538d0dc24bec3",
      "tx_id": "1a6742c537f2bb32473b01b1dcb1caac942722a4787bb185e4791916926e5fdb",
      "output_index": 8485,
      "tx_pointer_block_height": 758127325,
      "tx_pointer_tx_idx": 50052,
      "states": [
        {
          "key": "71bb3e4634561f8892e604021ab96b632eafa229ec3dd885442b44972526c4e8",
          "value": "ce25a0416c3955d818bb1ab0e832877d628233e3e1d64f76b71a80ec4f332f4d"
        }
      ],
      "balances": [
        {
          "asset_id": "2f736ab973e87f53b4857842789a1dd59167629b3c5cabe47dbf48827ec0818e",
          "amount": 2760603075432636416
        }
      ]
    },
    {
      "contract_id": "86cc533ae2e5151fa164130012dd3f09e8861767739da0c536da6da933caf851",
      "code": "4b297f047c3beda43d6e9f9b1b1bc44cdfce3574e7f7f0b2de2323a06ebc356b",
      "tx_id": "aa291871f4b9d120f76583a1da5d1123f2b8a0ac606b940a66e42efd782af520",
      "output_index": 561,
      "tx_pointer_block_height": 3943236114,
      "tx_pointer_tx_idx": 63027,
      "states": [
        {
          "key": "ad614726b313d811012d4be723755687120a7c6df2389c373c8fbe06bda7a611",
          "value": "20652668dd5687fee73fac7a2c564da689e9da76e1430d191fbc937d42f74104"
        }
      ],
      "balances": [
        {
          "asset_id": "1adc2ed6f402dca396bfe6dfa3ac8a57ea08860870fd515cf7ca6afb887dd1e0",
          "amount": 504295306752149147
        }
      ]
    }
  ],
  "last_block": {
    "block_height": 1958882788,
    "da_block_height": 7511963279655444049,
    "consensus_parameters_version": 3579832892,
    "state_transition_version": 1138407478,
    "blocks_root": "4076623c0231898be0722c08e4eb229c44cb957d29cf7370af8a1b9c542d49b3"
  }
}
